### PR TITLE
Input checking in af::sparse (Fixes #2134)

### LINKED ISSUE #2134: Input checking in af::sparse
This report is motivated by https://stackoverflow.com/questions/49867065/arrayfire-sparse-matrix-issues. In that SO post the OP used input in effectively COO format to create a sparse matrix in CSR format:
```
int row[] = {0,0,0,1,1,2,2};
int col[] = {0,1,2,0,1,0,2};
double values[] = { 3,3, 4,3,10,4,3};
array rr = sparse(3,3,array(7,values),array(7,row),array(7,col));
af_print(rr);
af_print(dense(rr));
```
One can fix this by either using the correct row indices:
```
int row[] = {0,3,5,7};
int col[] = {0,1,2,0,1,0,2};
float values[] = {3,3,4,3,10,4,3};
array rr = sparse(3,3,array(7,values),array(4,row),array(7,col));
af_print(rr);
af_print(dense(rr)); 
```
or the correct format:
```
int row[] = {0,0,0,1,1,2,2};
int col[] = {0,1,2,0,1,0,2};
float values[] = { 3,3, 4,3,10,4,3};
array rr = sparse(3,3,array(7,values),array(7,row),array(7,col), AF_STORAGE_COO);
af_print(rr);
af_print(dense(rr));
```
But why did the OP not get an error when trying to create a CSR matrix with incorrect input?

### PR DESCRIPTION
**Pull request by rstub**
Fixes #2134.

### CODE REVIEW

File: test/sparse.cpp [hunk 1]
```diff
@@ -119,6 +119,51 @@ TEST(Sparse, ISSUE_1745)
...
+  int cols[] = {0,1,2,0,1,0,2};
+  float values[] = {3,3,4,3,10,4,3};
+  af::array row(4, rows);
+  af::array col(7, cols);
+  af::array value(7, values);
+  af_array A = 0;
+  EXPECT_EQ(AF_SUCCESS, af_create_sparse_array(&A, 3, 3, value.get(), row.get(), col.get(), AF_STORAGE_CSR));
+  EXPECT_EQ(AF_ERR_SIZE, af_create_sparse_array(&A, 3, 3, value.get(), row.get(), col.get(), AF_STORAGE_CSC));
+  EXPECT_EQ(AF_ERR_SIZE, af_create_sparse_array(&A, 3, 3, value.get(), row.get(), col.get(), AF_STORAGE_COO));
+  if(A != 0) af_release_array(A);
```
Comment (pavanky) [inline test/sparse.cpp]: This line should theoretically be present after each `EXPECT_EQ`.

Comment (rstub) [reply test/sparse.cpp]: No problem. I will add that right away.

**Comment by umar456** [comment]
Looks like the tests are failing because of a segfault in Sparse.ISSUE_2134_CSC

http://ci.arrayfire.org/testDetails.php?test=224076&build=2255

**Comment by pavanky** [comment]
@rstub I think you need to reset `A` to `0` every time. `af_release_array` does not reset the value.

**Comment by rstub** [comment]
I cannot reproduce these segfaults locally, so htis a bit trial and error. Let's see if explicitly setting to 0 helps.

**Comment by pavanky** [comment]
@rstub looks like the tests are passing.

**Pull request merged**